### ISSUE TITLE
Cityscapes model and evaluation

### ISSUE THREAD

**Issue by vobecant** [janghyuncho/PiCIE#16]
Dear authors,

thank you very much for your work. I would be interested in using your model trained on Cityscapes, can you please provide it?
Also, I wonder how exactly do you evaluate your Cityscapes model. In the paper, you write
> After filtering out void group, we have 27 categories.

However, in the standard Cityscapes evaluation, only 19 classes are evaluated. Can you please explain how exactly do you do the evaluation?

Thank you very much in advance.

**Comment by janghyuncho** [comment]
Hi vobecant, 

Thank you a lot for your interest in our paper! And my apologies for the late reply, I have been very busy lately. 

First of all, [here](https://urldefense.com/v3/__https://drive.google.com/file/d/1phk0hlpFT2RMmOqzQ6Csoj6HgnZFskbK/view?usp=sharing__;!!OToaGQ!4f5kMMLjSy4HWv9ov9Th7sF3SGJ4ZfvKkeQjRLpqnPPoZxGh7Hu5Lcqhg6F-awOuR4lG$) is the pre-trained PiCIE on Cityscapes. The evaluation is done by setting "ignore" classes to -1, which effectively leaves 27 classes after all. 

I appreciate your patience and again, my apologies for the delayed answer.

**Comment by zyy-cn** [comment]
> Hi vobecant,
> 
> Thank you a lot for your interest in our paper! And my apologies for the late reply, I have been very busy lately.
> 
> First of all, [here](https://urldefense.com/v3/__https://drive.google.com/file/d/1phk0hlpFT2RMmOqzQ6Csoj6HgnZFskbK/view?usp=sharing__;!!OToaGQ!4f5kMMLjSy4HWv9ov9Th7sF3SGJ4ZfvKkeQjRLpqnPPoZxGh7Hu5Lcqhg6F-awOuR4lG$) is the pre-trained PiCIE on Cityscapes. The evaluation is done by setting "ignore" classes to -1, which effectively leaves 27 classes after all.
> 
> I appreciate your patience and again, my apologies for the delayed answer.

Do you mean setting 'void' groups to -1 (as described in your paper) based the definition [here](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/helpers/labels.py)? If do so, it will resulting in exactly 27 classes. However if set 'ignore' to -1, it would leaves 19 classes.

**Comment by janghyuncho** [comment]
Hi zyy-cn,

Yes, you are absolutely right! I now understand where the confusion came from. We ignored the categories that are vague in semantic meaning (hence, "void"). My apologies for the confusion that it might have caused.

**Comment by vobecant** [comment]
Hi @janghyuncho ,
thank you very much for your reply. However, when I relabel the labels in a way that you have just described (labels from 0 to 6 are set to _ignore_, and the rest are shifted to be zero-based), I get the results mIoU ~6.9 and pixel accuracy ~50.5
I wonder what I am doing wrong. I predict on the full-scale images, normalize them by `mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]`, and the do:
```
with torch.no_grad():
    for i, (image, label) in enumerate(tqdm(dataloader)):
        image = image.cuda()
        feats = model(image)

        # label = relabel_cs(label)
        label = relabel_cs_picie(label)

        if args.metric_test == 'cosine':
            feats = F.normalize(feats, dim=1, p=2)

        B, C, H, W = feats.size()
        if i == 0:
            logger.info('Batch input size   : {}'.format(list(image.shape)))
            logger.info('Batch label size   : {}'.format(list(label.shape)))
            logger.info('Batch feature size : {}\n'.format(list(feats.shape)))

        probs = classifier(feats)
        probs = F.interpolate(probs, label.shape[-2:], mode='bilinear', align_corners=False)
        preds = probs.topk(1, dim=1)[1].view(B, -1).cpu().numpy()
        label = label.view(B, -1).cpu().numpy()

        histogram += scores(label, preds, args.K_test)

        if i % 20 == 0:
            logger.info('{}/{}'.format(i, len(dataloader)))

# Hungarian Matching.
m = linear_assignment(histogram.max() - histogram)
m = np.concatenate((m[0][:, None], m[1][:, None]), axis=1)

# Evaluate.
acc = histogram[m[:, 0], m[:, 1]].sum() / histogram.sum() * 100

new_hist = np.zeros((args.K_test, args.K_test))
for idx in range(args.K_test):
    new_hist[m[idx, 1]] = histogram[idx]

# NOTE: Now [new_hist] is re-ordered to 12 thing + 15 stuff classses.
res1 = get_result_metrics(new_hist)
logger.info('ACC  - All: {:.4f}'.format(res1['overall_precision (pixel accuracy)']))
logger.info('mIOU - All: {:.4f}'.format(res1['mean_iou']))
```
Can you please give me a hint?

Thank you very much in advance.

**Comment by janghyuncho** [comment]
Hi, I uploaded the training end evaluation code for cityscapes for [picie](https://github.com/janghyuncho/PiCIE/blob/master/train_picie.py) and [mdc](https://github.com/janghyuncho/PiCIE/blob/master/train_mdc.py) ([log](https://github.com/janghyuncho/PiCIE/tree/master/logs/reproduce)). 

Sorry for the delay. 

Thank you!

**Comment by vobecant** [comment]
Hi, thank you very much for sharing the codes! If I understand the code correctly, you do the center cropping even for the testing dataset, right?
To be precise, I run the following code:
```
python train_picie.py --data_root datasets/cityscapes/ --save_root results/cityscapes/picie/train/1/ --pretrain --lr 0.001 --repeats 1 --seed 1 --num_init_batches 2
0 --num_batches 1 --kmeans_n_iter 20 --K_train 27 --K_test 27 --batch_size_cluster 256 --num_epoch 10 --cityscapes --label_mode 'gtFine' --res 320 --res1 320 --res2 640 --augment --grey --blur --jitter --
equiv --random_crop --h_flip --eval_only --eval_path ./trained_models/picie_city.tar --num_workers 0
```
But I get the following results, which are different to the results from the paper:
```

Is there anyone who managed to get the results reported in the paper? Or at least something close?
Thank you.
ACC  - All: 65.5020
mIOU - All: 11.4750
```

**Comment by janghyuncho** [comment]
Hi vobecant,

The reported result was with center-crop during testing. In our experiment, about 2 points of deviation in accuracy and 1 point in mIoU are expected due to the nature of unsupervised learning setting. This is a limitation in our work but we did see similar if not greater deviations in other methods as well. 

You can test without center crop by setting "--long_image" argument in the command. Below is what I got:
`
ACC  - All: 63.3444
mIOU - All: 11.2635
`
Hope this answers your question and thank you again for your interest in our paper!
